File: CS/CustomLogonParametersExample.Module/CustomLogonParameters.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.ComponentModel;
using DevExpress.Xpo;
using DevExpress.Data.Filtering;
using DevExpress.Persistent.Base;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Utils;
using DevExpress.ExpressApp.Security;
using DevExpress.ExpressApp.Model;
using CustomLogonParametersExample.Module.BusinessObjects;


namespace CustomLogonParametersExample.Module {
    [NonPersistent, ModelDefault("Caption", "Log On")]
    public class CustomLogonParameters : ICustomObjectSerialize, INotifyPropertyChanged {
        public event PropertyChangedEventHandler PropertyChanged;
        private void RaisePropertyChanged(string name) {
            if (PropertyChanged == null) return;
            PropertyChanged(this, new PropertyChangedEventArgs(name));
        }

        private Company company;
        [DataSourceProperty("AvailableCompanies"), ImmediatePostData]
        public Company Company { 
            get { return company; }
            set {
                if (company == value) return;
                company = value;
                RaisePropertyChanged("Company");
                RefreshAvailableUsers();
            }
        }
        private Employee employee;
        [DataSourceProperty("AvailableUsers"), ImmediatePostData]
        public Employee Employee {
            get { return employee; }
            set {
                if (employee == value) return;
                employee = value;
                if (employee != null) {
                    UserName = employee.UserName;
                    if (Company == null) Company = employee.Company;
                }
                else UserName = string.Empty;
                RaisePropertyChanged("Employee");
            }
        }
        private void RefreshAvailableUsers() {
            if (availableUsers == null) {
                return;
            }
            if (Company == null) availableUsers.Criteria = null;
            else availableUsers.Criteria = 
                new BinaryOperator("Company", Company);
            if (employee != null && availableUsers.IndexOf(employee) == -1) {
                Employee = null;
            }
        }
        private string password;
        [PasswordPropertyText(true)]
        public string Password {
            get { return password; }
            set {
                if (password == value) return;
                password = value;
                RaisePropertyChanged("Password");
            }
        }
        private IObjectSpace objectSpace;
        private XPCollection<Company> availableCompanies;
        private XPCollection<Employee> availableUsers;
        [Browsable(false)]
        public IObjectSpace ObjectSpace {
            get { return objectSpace; }
            set { objectSpace = value; }
        }
        [Browsable(false)]
        [CollectionOperationSet(AllowAdd = false)]
        public XPCollection<Company> AvailableCompanies {
            get {
                if (availableCompanies == null) {
                    availableCompanies = ObjectSpace.GetObjects<Company>() as XPCollection<Company>;
                }
                return availableCompanies;
            }
        }
        [Browsable(false)]
        [CollectionOperationSet(AllowAdd = false)]
        public XPCollection<Employee> AvailableUsers {
            get {
                if (availableUsers == null) {
                    availableUsers = ObjectSpace.GetObjects<Employee>() as XPCollection<Employee>;
                    RefreshAvailableUsers();
                }
                return availableUsers;
            }
        }
        public void Reset() {
            ObjectSpace = null;
            availableCompanies = null;
            availableUsers = null;
            Company = null;
            Employee = null;
            Password = null;
        }
        [Browsable(false)]
        public string UserName { get; set; }
        public void ReadPropertyValues(SettingsStorage storage) {
            Employee = objectSpace.FindObject<Employee>(
                new BinaryOperator("UserName", storage.LoadOption("", "UserName")));
            if (Employee != null) Company = Employee.Company;
        }
        public void WritePropertyValues(SettingsStorage storage) {
            storage.SaveOption("", "UserName", UserName);
        }
    }
}

File: VB/CustomLogonParametersExample.Module/CustomLogonParameters.vb
Imports System.Text
Imports System.ComponentModel
Imports DevExpress.Xpo
Imports DevExpress.Data.Filtering
Imports DevExpress.Persistent.Base
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Utils
Imports DevExpress.ExpressApp.Security
Imports DevExpress.ExpressApp.Model
Imports CustomLogonParametersExample.Module.BusinessObjects


Namespace CustomLogonParametersExample.Module
    <NonPersistent, ModelDefault("Caption", "Log On")> _
    Public Class CustomLogonParameters
        Implements ICustomObjectSerialize

        Private _company As Company
        <DataSourceProperty("AvailableCompanies"), ImmediatePostData> _
        Public Property Company() As Company
            Get
                Return _company
            End Get
            Set(ByVal value As Company)
                _company = value
                RefreshAvailableUsers()
            End Set
        End Property
        Private _employee As Employee
        <DataSourceProperty("AvailableUsers"), ImmediatePostData> _
        Public Property Employee() As Employee
            Get
                Return _employee
            End Get
            Set(ByVal value As Employee)
                _employee = value
                If _employee IsNot Nothing Then
                    UserName = _employee.UserName
                    If Company Is Nothing Then
                        Company = _employee.Company
                    End If
                Else
                    UserName = String.Empty
                End If

            End Set
        End Property
        Private Sub RefreshAvailableUsers()
            If _availableUsers Is Nothing Then
                Return
            End If
            If Company Is Nothing Then
                _availableUsers.Criteria = Nothing
            Else
                _availableUsers.Criteria = New BinaryOperator("Company", Company)
            End If
            If _employee IsNot Nothing AndAlso _availableUsers.IndexOf(_employee) = -1 Then
                Employee = Nothing
            End If
        End Sub
        Private privatePassword As String
        <PasswordPropertyText(True)> _
        Public Property Password() As String
            Get
                Return privatePassword
            End Get
            Set(ByVal value As String)
                privatePassword = value
            End Set
        End Property
        Private _objectSpace As IObjectSpace
        Private _availableCompanies As XPCollection(Of Company)
        Private _availableUsers As XPCollection(Of Employee)
        <Browsable(False)> _
        Public Property ObjectSpace() As IObjectSpace
            Get
                Return _objectSpace
            End Get
            Set(ByVal value As IObjectSpace)
                _objectSpace = value
            End Set
        End Property
        <Browsable(False)> _
        Public ReadOnly Property AvailableCompanies() As XPCollection(Of Company)
            Get
                If _availableCompanies Is Nothing Then
                    _availableCompanies = TryCast(ObjectSpace.GetObjects(Of Company)(), XPCollection(Of Company))
                    _availableCompanies.BindingBehavior = CollectionBindingBehavior.AllowNone
                End If
                Return _availableCompanies
            End Get
        End Property
        <Browsable(False)> _
        Public ReadOnly Property AvailableUsers() As XPCollection(Of Employee)
            Get
                If _availableUsers Is Nothing Then
                    _availableUsers = TryCast(ObjectSpace.GetObjects(Of Employee)(), XPCollection(Of Employee))
                    _availableUsers.BindingBehavior = CollectionBindingBehavior.AllowNone
                    RefreshAvailableUsers()
                End If
                Return _availableUsers
            End Get
        End Property
        Public Sub Reset()
            ObjectSpace = Nothing
            _availableCompanies = Nothing
            _availableUsers = Nothing
            Company = Nothing
            Employee = Nothing
            Password = Nothing
        End Sub
        Private privateUserName As String
        <Browsable(False)> _
        Public Property UserName() As String
            Get
                Return privateUserName
            End Get
            Set(ByVal value As String)
                privateUserName = value
            End Set
        End Property
        Public Sub ReadPropertyValues(ByVal storage As SettingsStorage) Implements ICustomObjectSerialize.ReadPropertyValues
            Employee = _objectSpace.FindObject(Of Employee)(New BinaryOperator("UserName", storage.LoadOption("", "UserName")))
            If Employee IsNot Nothing Then
                Company = Me.Employee.Company
            End If
        End Sub
        Public Sub WritePropertyValues(ByVal storage As SettingsStorage) Implements ICustomObjectSerialize.WritePropertyValues
            storage.SaveOption("", "UserName", UserName)
        End Sub
    End Class
End Namespace
